File: app/src/main/java/org/scp/gymlog/ui/exercises/ExercisesActivity.java
package org.scp.gymlog.ui.exercises;

import static org.scp.gymlog.util.Constants.INTENT;
import static org.scp.gymlog.util.LambdaUtils.valueEquals;

import android.content.Intent;
import android.content.SharedPreferences;
import android.os.Bundle;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;

import androidx.preference.PreferenceManager;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import org.scp.gymlog.R;
import org.scp.gymlog.exceptions.InternalException;
import org.scp.gymlog.model.Exercise;
import org.scp.gymlog.model.Muscle;
import org.scp.gymlog.model.Order;
import org.scp.gymlog.room.AppDatabase;
import org.scp.gymlog.room.DBThread;
import org.scp.gymlog.ui.common.DBAppCompatActivity;
import org.scp.gymlog.ui.common.components.TrainingFloatingActionButton;
import org.scp.gymlog.ui.common.dialogs.TextDialogFragment;
import org.scp.gymlog.ui.createexercise.CreateExerciseActivity;
import org.scp.gymlog.ui.registry.RegistryActivity;
import org.scp.gymlog.util.Data;

import java.util.List;

public class ExercisesActivity extends DBAppCompatActivity {
    private int muscleId;
    private List<Integer> exercisesId;
    private Order order;
    private ExercisesRecyclerViewAdapter recyclerAdapter;
    private RecyclerView exercisesRecyclerView;

    @Override
    protected int onLoad(Bundle savedInstanceState, AppDatabase db) {
        muscleId = getIntent().getExtras().getInt("muscleId");
        exercisesId = db.exerciseDao().getExercisesIdByMuscleId(muscleId);
        return CONTINUE;
    }

    @Override
    protected void onDelayedCreate(Bundle savedInstanceState) {
        setContentView(R.layout.activity_exercises);

        SharedPreferences preferences = PreferenceManager.getDefaultSharedPreferences(this);
        order = Order.getByName(
                preferences.getString("exercisesSortLastUsed", Order.ALPHABETICALLY.name));

        Muscle muscle = Data.getInstance().getMuscles().stream()
                .filter(gr -> gr.getId() == muscleId)
                .findFirst()
                .orElseThrow(() -> new InternalException("Muscle id not found"));

        setTitle(muscle.getText());
        exercisesRecyclerView = findViewById(R.id.exercisesList);
        exercisesRecyclerView.setLayoutManager(new LinearLayoutManager(this));
        recyclerAdapter = new ExercisesRecyclerViewAdapter(exercisesId, this, order,
                this::onExerciseItemMenuSelected);
        recyclerAdapter.setOnClickListener(ex -> {
            Intent intent = new Intent(this, RegistryActivity.class);
            intent.putExtra("exerciseId", ex.getId());
            startActivityForResult(INTENT.REGISTRY, intent);
        });
        exercisesRecyclerView.setAdapter(recyclerAdapter);

        TrainingFloatingActionButton fab = findViewById(R.id.fabTraining);
        fab.updateFloatingActionButton();
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        MenuInflater inflater = getMenuInflater();
        inflater.inflate(R.menu.exercises_menu, menu);

        switch (order) {
            case ALPHABETICALLY: menu.findItem(R.id.sortAlphabetically).setChecked(true); break;
            case LAST_USED:      menu.findItem(R.id.sortLastUsed).setChecked(true);       break;
        }
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        if (item.getItemId() == R.id.create_button) {
            Intent intent = new Intent(this, CreateExerciseActivity.class);
            intent.putExtra("muscleId", muscleId);
            startActivityForResult(INTENT.CREATE_EXERCISE_FROM_MUSCLE, intent);

        } else if (item.getItemId() == R.id.sortAlphabetically) {
            recyclerAdapter.switchOrder(order = Order.ALPHABETICALLY);
            SharedPreferences preferences = PreferenceManager.getDefaultSharedPreferences(this);
            SharedPreferences.Editor editor = preferences.edit();
            editor.putString("exercisesSortLastUsed", order.name);
            editor.apply();
            item.setChecked(true);

        } else if (item.getItemId() == R.id.sortLastUsed) {
            recyclerAdapter.switchOrder(order = Order.LAST_USED);
            SharedPreferences preferences = PreferenceManager.getDefaultSharedPreferences(this);
            SharedPreferences.Editor editor = preferences.edit();
            editor.putString("exercisesSortLastUsed", order.name);
            editor.apply();
            item.setChecked(true);
        }

        return false;
    }

    private void onExerciseItemMenuSelected(Exercise exercise, int action) {
        if (action == R.id.editExercise) {
            Intent intent = new Intent(this, CreateExerciseActivity.class);
            intent.putExtra("exerciseId", exercise.getId());
            startActivityForResult(INTENT.EDIT_EXERCISE, intent);

        } else if (action == R.id.removeExercise) {
            TextDialogFragment dialog = new TextDialogFragment(R.string.dialog_confirm_remove_exercise_title,
                    R.string.dialog_confirm_remove_exercise_text,
                    confirmed -> {
                        if (confirmed) {
                            DBThread.run(this, db -> {
                                if (db.exerciseDao().delete(exercise.toEntity()) == 1) {
                                    runOnUiThread(() -> recyclerAdapter.removeExercise(exercise));
                                    db.trainingDao().deleteEmptyTraining();

                                    exercisesId.remove((Integer) exercise.getId());
                                    Data.getInstance().getExercises().removeIf(valueEquals(exercise));
                                }
                            });

                        }
                    });
            dialog.show(getSupportFragmentManager(), null);
        }
    }

    @Override
    public void onActivityResult(int intentResultId, Intent data) {
        if (intentResultId == INTENT.EDIT_EXERCISE) {
            int id = data.getIntExtra("exerciseId", -1);
            Exercise ex = Data.getExercise(id);
            boolean hasMuscle = ex.getPrimaryMuscles().stream().map(Muscle::getId)
                    .anyMatch(valueEquals(muscleId));
            if (hasMuscle) {
                recyclerAdapter.updateNotify(ex);
            } else {
                recyclerAdapter.removeExercise(ex);
                exercisesId.removeIf(valueEquals(ex.getId()));
            }

        } else if (intentResultId == INTENT.CREATE_EXERCISE_FROM_MUSCLE) {
            int id = data.getIntExtra("exerciseId", -1);
            Exercise ex = Data.getExercise(id);
            boolean hasMuscle = ex.getPrimaryMuscles().stream().map(Muscle::getId)
                    .anyMatch(valueEquals(muscleId));
            if (hasMuscle) {
                exercisesId.add(ex.getId());
                recyclerAdapter.addExercise(ex);
            }
        } else if (intentResultId == INTENT.REGISTRY) {
            if (data.getBooleanExtra("refresh", false)) {
                if (order.equals(Order.ALPHABETICALLY)) {
                    int id = data.getIntExtra("exerciseId", -1);
                    Exercise ex = Data.getExercise(id);
                    recyclerAdapter.updateNotify(ex);

                } else {
                    exercisesRecyclerView.scrollToPosition(0);
                    recyclerAdapter.switchOrder(order);
                }
            }
        }
    }
}

File: app/src/main/java/org/scp/gymlog/ui/exercises/ExercisesActivity.kt
package org.scp.gymlog.ui.exercises

import android.content.Intent
import android.os.Bundle
import android.view.Menu
import android.view.MenuItem
import org.scp.gymlog.R
import org.scp.gymlog.databinding.ListitemExercisesRowBinding
import org.scp.gymlog.model.Exercise
import org.scp.gymlog.model.Muscle
import org.scp.gymlog.model.Order
import org.scp.gymlog.model.Order.Companion.getByCode
import org.scp.gymlog.model.Variation
import org.scp.gymlog.ui.common.CustomAppCompatActivity
import org.scp.gymlog.ui.common.components.TrainingFloatingActionButton
import org.scp.gymlog.ui.common.components.listView.SimpleListView
import org.scp.gymlog.ui.common.dialogs.MenuDialogFragment
import org.scp.gymlog.ui.common.dialogs.TextDialogFragment
import org.scp.gymlog.ui.create.CreateExerciseActivity
import org.scp.gymlog.ui.preferences.PreferencesDefinition
import org.scp.gymlog.ui.registry.RegistryActivity
import org.scp.gymlog.ui.top.TopActivity
import org.scp.gymlog.util.Constants.IntentReference
import org.scp.gymlog.util.Data
import org.scp.gymlog.util.extensions.DatabaseExts.dbThread
import org.scp.gymlog.util.extensions.MessagingExts.toast
import org.scp.gymlog.util.extensions.PreferencesExts.loadString
import org.scp.gymlog.util.extensions.PreferencesExts.save
import org.scp.gymlog.util.extensions.RedirectionExts.goToVariation

class ExercisesActivity : CustomAppCompatActivity() {

    private lateinit var muscle: Muscle
    private lateinit var exercises: MutableList<Exercise>

    private var trainingFloatingButton: TrainingFloatingActionButton? = null
    private lateinit var exercisesListView: SimpleListView<Exercise, ListitemExercisesRowBinding>
    private lateinit var handler: ExercisesListHandler

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        val muscleId = intent.extras!!.getInt("muscleId")
        muscle = Data.getMuscle(muscleId)

        exercises = Data.exercises
            .filter { it.primaryMuscles.contains(muscle) }
            .toMutableList()

        setContentView(R.layout.activity_exercises)

        setTitle(muscle.text)

        exercisesListView = findViewById(R.id.exercisesList)

        handler = ExercisesListHandler(this, exercisesListView, muscle)
        exercisesListView.init(exercises, handler)
        handler.init()
        handler.onExerciseClicked(this::itemClicked)
        handler.onVariationClicked(this::itemClicked)

        val expandExerciseId = intent.extras!!.getInt("expandExerciseId", -1)
        if (expandExerciseId > 0) {
            val expandExercise = exercises.find { it.id == expandExerciseId }
            if (expandExercise != null && expandExercise.gymVariations.size > 1) {
                exercisesListView.updateState(expandExercise) {
                    it["expanded"] = true
                    true
                }
            }
        }

        trainingFloatingButton = findViewById<TrainingFloatingActionButton>(R.id.fabTraining).apply {
            updateFloatingActionButton()
        }
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        val inflater = menuInflater
        inflater.inflate(R.menu.exercises_menu, menu)

        when (getByCode(loadString(PreferencesDefinition.EXERCISES_ORDER))) {
            Order.ALPHABETICALLY -> menu.findItem(R.id.sortAlphabetically).isChecked = true
            Order.LAST_USED -> menu.findItem(R.id.sortLastUsed).isChecked = true
        }
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.searchButton -> {
                val intent = Intent(this, SearchActivity::class.java)
                intent.putExtra("muscleId", muscle.id)
                startActivityForResult(intent, IntentReference.SEARCH_LIST)
            }
            R.id.create_button -> {
                val intent = Intent(this, CreateExerciseActivity::class.java)
                intent.putExtra("muscleId", muscle.id)
                startActivityForResult(intent, IntentReference.CREATE_EXERCISE_FROM_MUSCLE)
            }
            R.id.latestButton -> {
                Data.training?.id
                    ?.also { trainingId ->
                        dbThread { db ->
                            db.bitDao()
                                .getMostRecentByTrainingId(trainingId)
                                ?.let { Data.getVariation(it.variationId) }
                                ?.let { goToVariation(it) }
                                ?: toast(R.string.validation_no_exercise_registered)
                        }
                    }
                    ?: toast(R.string.validation_training_not_started)
            }
            R.id.sortAlphabetically -> {
                val order = Order.ALPHABETICALLY
                handler.updateOrder(order)
                save(PreferencesDefinition.EXERCISES_ORDER, order.code)
                item.isChecked = true
            }
            R.id.sortLastUsed -> {
                val order = Order.LAST_USED
                handler.updateOrder(order)
                save(PreferencesDefinition.EXERCISES_ORDER, order.code)
                item.isChecked = true
            }
        }
        return false
    }

    private fun itemClicked(exercise: Exercise, long: Boolean) {
        if (long) {
            val deletionPref = loadString(PreferencesDefinition.EXERCISES_DELETION)
            val removedItems = if (deletionPref == "0") listOf(R.id.removeExercise) else listOf()

            MenuDialogFragment(R.menu.exercise_menu, removedItems) { action ->
                exerciseMenuActionSelected(exercise, action)
            }.apply { show(supportFragmentManager, null) }
        } else {
            val variation = exercise.variations.first { it.default }
            val intent = Intent(this, RegistryActivity::class.java)
            intent.putExtra("exerciseId", exercise.id)
            intent.putExtra("variationId", variation.id)
            startActivityForResult(intent, IntentReference.REGISTRY)
        }
    }

    private fun itemClicked(variation: Variation) {
        val exercise = variation.exercise
        val intent = Intent(this, RegistryActivity::class.java)
        intent.putExtra("exerciseId", exercise.id)
        intent.putExtra("variationId", variation.id)
        startActivityForResult(intent, IntentReference.REGISTRY)
    }

    private fun exerciseMenuActionSelected(exercise: Exercise, action: Int) {
        when (action) {
            R.id.open -> {
                itemClicked(exercise, false)
            }
            R.id.topRanking -> {
                val intent = Intent(this, TopActivity::class.java)
                intent.putExtra("exerciseId", exercise.id)
                startActivity(intent)
            }
            R.id.editExercise -> {
                val intent = Intent(this, CreateExerciseActivity::class.java)
                intent.putExtra("exerciseId", exercise.id)
                startActivityForResult(intent, IntentReference.EDIT_EXERCISE)
            }
            R.id.removeExercise -> {
                val dialog = TextDialogFragment(
                    R.string.dialog_confirm_remove_exercise_title,
                    R.string.dialog_confirm_remove_exercise_text
                ) { confirmed ->
                    if (confirmed) {
                        dbThread { db ->
                            if (db.exerciseDao().delete(exercise.toEntity()) == 1) {
                                runOnUiThread { exercisesListView.remove(exercise) }
                                db.trainingDao().deleteEmptyTraining()
                                exercises.remove(exercise)
                                Data.exercises.removeIf { it === exercise }
                            }
                        }
                    }
                }
                dialog.show(supportFragmentManager, null)
            }
        }
    }

    override fun onResume() {
        super.onResume()
        trainingFloatingButton?.updateFloatingActionButton()
    }

    override fun onActivityResult(intentReference: IntentReference, data: Intent) {
        when (intentReference) {
            IntentReference.EDIT_EXERCISE -> {
                val exerciseId = data.getIntExtra("exerciseId", -1)
                val exercise = Data.getExercise(exerciseId)
                val hasMuscle = exercise.primaryMuscles
                    .any { it.id == muscle.id }

                if (hasMuscle) {
                    exercisesListView.notifyUpdate(exercise)
                } else {
                    exercisesListView.remove(exercise)
                    exercises.remove(exercise)
                }
            }
            IntentReference.CREATE_EXERCISE_FROM_MUSCLE -> {
                val exerciseId = data.getIntExtra("exerciseId", -1)
                val exercise = Data.getExercise(exerciseId)
                val hasMuscle = exercise.primaryMuscles
                    .any { it.id == muscle.id }

                if (hasMuscle) {
                    exercises.add(exercise)
                    exercisesListView.add(exercise)
                }
            }
            IntentReference.SEARCH_LIST -> {
                if (data.getBooleanExtra("refresh", false)) {
                    exercises = Data.exercises
                        .filter { it.primaryMuscles.contains(muscle) }
                        .toMutableList()

                    exercisesListView.setListData(exercises)
                }
            }
            IntentReference.REGISTRY -> {
                if (data.getBooleanExtra("refresh", false)) {
                    val order = getByCode(loadString(PreferencesDefinition.EXERCISES_ORDER))

                    if (order == Order.ALPHABETICALLY) {
                        val id = data.getIntExtra("exerciseId", -1)
                        val ex = Data.getExercise(id)
                        exercisesListView.notifyUpdate(ex)
                    } else {
                        exercisesListView.forceReorder()
                    }
                }
            }
            else -> {}
        }
    }
}